### PR TITLE
Implement movd, psrlq, movdqa

### PR DESCRIPTION
**Pull request by MatthewMerrill**
Startblock:
 #533 is submitted

### CODE REVIEW

File: emu/decode.h [hunk 1]
```diff
@@ -107,6 +107,30 @@ __no_instrument DECODER_RET glue(DECODER_NAME, OP_SIZE)(DECODER_ARGS) {
...
+                           // TODO: REX.W = 1 might be needed later
+                           READMODRM; VZLOAD(modrm_val, xmm_modrm_reg,32);
+                           break;
+
+                case 0x6f: TRACEI("movdqa xmm:modrm, xmm");
+                           READMODRM; VLOAD(xmm_modrm_val, xmm_modrm_reg,128); break;
+
+                case 0x73: READMODRM;
+                           switch (modrm.opcode) {
+                               case (0x02): TRACEI("psrlq xmm imm");
```
[the remark this answers is not in the document]
Comment (MatthewMerrill) [reply emu/decode.h]: Did you hear about the two detectives writing up a dissertation?

[the remark this answers is not in the document]
Comment (MatthewMerrill) [reply emu/decode.h]: case pair in thesis writing

[the remark this answers is not in the document]
Comment (tbodt) [reply emu/decode.h]: no

File: emu/decode.h [hunk 2]
```diff
@@ -107,6 +107,30 @@ __no_instrument DECODER_RET glue(DECODER_NAME, OP_SIZE)(DECODER_ARGS) {
...
+                                            READIMM8; VIMM_SHIFTR(xmm_modrm_val, imm,64); break;
+                               default: UNDEFINED;
+                           }
+                           break;
+
+                case 0x7e: TRACEI("movd xmm, modrm");
+                           // TODO: REX.W = 1 might be needed later
```
[the remark this answers is not in the document]
Comment (MatthewMerrill) [reply emu/decode.h]: image

File: emu/decode.h [hunk 3]
```diff
@@ -107,6 +107,30 @@ __no_instrument DECODER_RET glue(DECODER_NAME, OP_SIZE)(DECODER_ARGS) {
...
 
+                case 0x6e: TRACEI("movd modrm, xmm");
+                           // TODO: REX.W = 1 might be needed later
+                           READMODRM; VZLOAD(modrm_val, xmm_modrm_reg,32);
+                           break;
+
+                case 0x6f: TRACEI("movdqa xmm:modrm, xmm");
+                           READMODRM; VLOAD(xmm_modrm_val, xmm_modrm_reg,128); break;
+
+                case 0x73: READMODRM;
```
Comment (tbodt) [inline emu/decode.h]: Please put the READMODRM inside the switch, there might be opcodes that don't have a modrm

File: emu/sse.c [hunk 4]
```diff
@@ -69,6 +69,15 @@ void vec_store128(struct cpu_state *UNUSED(cpu), union xmm_reg *dst, const union
     *dst = *src;
 }
 
+void vec_imm_shiftr64(struct cpu_state *UNUSED(cpu), const uint8_t amount, union xmm_reg *src) {
+    if ((amount & 0x7F) > 63) {
```
Comment (tbodt) [inline emu/sse.c]: Why the mask?

File: emu/sse.h [hunk 5]
```diff
@@ -32,6 +37,8 @@ void vec_store32(struct cpu_state *UNUSED(cpu), union xmm_reg *src, const union
 void vec_store64(struct cpu_state *UNUSED(cpu), union xmm_reg *src, const union xmm_reg *dst);
 void vec_store128(struct cpu_state *UNUSED(cpu), union xmm_reg *src, const union xmm_reg *dst);
 
+// TODO is uint right?
```
Comment (tbodt) [inline emu/sse.h]: uint8_t is right

File: emu/decode.h (hunk 3 above)
Comment (tbodt) [reply emu/decode.h]: nvm.

Comment (MatthewMerrill) [reply emu/decode.h]: no

File: jit/gen.c [hunk 6]
```diff
@@ -449,35 +451,55 @@ static inline bool gen_vec(enum arg rm, enum arg reg, void (*helper)(), gadget_t
...
     }
     return true;
 }
 
 #define _v(src, dst, helper, helper_gadgets, z) do { \
     extern gadget_t helper_gadgets[vec_arg_count]; \
-    if (!gen_vec(src, dst, helper, &helper_gadgets, state, &modrm, saved_ip, seg_gs)) return false; \
+    if (!gen_vec(src, dst, helper, &helper_gadgets, state, &modrm, 0, saved_ip, seg_gs)) return false; \
+} while (0)
+#define _vi(imm, dst, helper, helper_gadgets, z) do { \
```
Comment (tbodt) [inline jit/gen.c]: can you make this _v_imm too. pretty sure that's the last nit

Comment (MatthewMerrill) [reply jit/gen.c]: done

**Pull request merged**